modules/fs2/src/main/scala/com/snowplowanalytics/snowplow/enrich/fs2/io/Metrics.scala

```diff
@@ -64,21 +66,24 @@ object Metrics {
    * Technically `Resource` doesn't give us much as we don't allocate a thread pool,
    * but it will make sure the last report is issued
    */
-  def resource[F[_]: Sync]: Resource[F, Metrics[F]] =
-    Resource
-      .make(init) { case (res, _) => Sync[F].delay(res.close()) }
-      .map { case (res, reg) => make[F](res, reg) }
-
-  /** Initialise backend resources */
-  def init[F[_]: Sync]: F[(Slf4jReporter, MetricRegistry)] =
-    Sync[F].delay {
-      val registry = new MetricRegistry()
-      val logger = LoggerFactory.getLogger(LoggerName)
-      val reporter = Slf4jReporter.forRegistry(registry).outputTo(logger).build()
-      (reporter, registry)
+  def resource[F[_]: Sync](config: MetricsReporter): Resource[F, Metrics[F]] =
+    for {
+      registry <- Resource.liftF(Sync[F].delay((new MetricRegistry())))
+      rep <- reporter(config, registry)
+    } yield make(rep, registry)
+
+  def reporter[F[_]: Sync](config: MetricsReporter, registry: MetricRegistry): Resource[F, ScheduledReporter] =
+    config match {
+      case MetricsReporter.Stdout(_) =>
+        for {
+          logger <- Resource.liftF(Sync[F].delay(LoggerFactory.getLogger(LoggerName)))
+          reporter <- Resource.fromAutoCloseable(Sync[F].delay(Slf4jReporter.forRegistry(registry).outputTo(logger).build))
+        } yield reporter
+      case statsd: MetricsReporter.StatsD =>
+        StatsDReporter.resource(statsd, registry)
     }
 
-  def make[F[_]: Sync](reporter: Slf4jReporter, registry: MetricRegistry): Metrics[F] =
+  def make[F[_]: Sync](reporter: ScheduledReporter, registry: MetricRegistry): Metrics[F] =
```

Comment:
But `ScheduledReporter` reports on its own in a background thread IIRC? So we can remove the `report` method?

Reply:
A `ScheduledReporter` would normally run on its own thread.  But we choose to run it on our own thread pool using the `report` method.

This is not a new change: the old `Sf4jReporter` is a subclass of `ScheduledReporter` - so we already had this setup.

What's different now.... the StatsD reporter uses a StatsD client underneath, and that client runs on its own thread.

I think this is a good overall setup - the client's blocking operations are run on a dedicated thread, but we do not waste an extra thread for our reporters which are non-blocking.